src/plugins/lexer/mod.ts

```diff
@@ -0,0 +1,390 @@
+import { Mode, Token, Trace } from "./types/mod.ts";
+
+/*
+note:
+  lexer uses zero-based numbering for tracking
+  its position and the position of tokens.
+*/
+export class Lexer {
+  source: string = "";
+  // lexer position on file
+  position: number = -1;
+  // virtual position; used for tracking token positions.
+  virtual_position: number = -1;
+  // keeps track of elementary debug info useful for parser
+  trace: Trace = {
+    line: 1,
+    column: [0, 0],
+  };
+  // stores current built token
+  token: string = "";
+  // stores current lexical tokens
+  tokens: Token[] = [];
+  // stores mode lexer is interpreting in
+  mode: Mode = Mode.Token;
+
+  // optional initiator for .mtl source
+  constructor(input?: string) {
+    if (input !== undefined) {
+      this.source = input.replace(/(\r\n|\n|\r)/gm, "&eol") + "&eof";
+    }
+  }
+
+  // loads new .mtl source, avoiding reinitializing class
+  load(input: string) {
+    this.source = input.replace(/(\r\n|\n|\r)/gm, "&eol") + "&eof";
+    return this;
+  }
+
+  // starts generation of lexical tokens
+  get generate() {
+    this.position = -1;
+    this.virtual_position = -1;
+    this.token = "";
+    this.trace = {
+      line: 1,
+      column: [0, 0],
+    };
+    this.tokens = [];
+    this.mode = Mode.Token;
+
+    // basic character loop
+    while ((this.source.length) > this.position) {
+      this.comprehend(this.next());
+    }
+    return this.tokens;
+  }
+
+  // used to build an understanding of individual tokens through individual characters
+  comprehend(character: string): void {
+    // if comprehending new token, store initial `character` location for `trace`
+    if (
+      (this.token.length === 0) && (this.mode !== Mode.String) &&
+      (this.mode !== Mode.Compiler)
+    ) {
+      this.trace.column[0] = this.virtual_position;
+    }
+    // verify mode lexer is in
+    switch (this.mode) {
+      case Mode.Token: {
+        // switch to string mode
+        if ((character === '"') && (this.token.length === 0)) {
+          this.mode = Mode.String;
+          break;
+
+          // presume token is finished; generate
+        } else if (character === " ") {
+          if (this.token.length === 0) return;
+          this.checkToken();
+
+          // switch to compiler call mode
+        } else if (character === "&") {
+          this.mode = Mode.Compiler;
+          this.checkToken();
+
+          // check if current character is an operator
+        } else if (this.isOperator(character) === true) {
+          // handles operator being beside another token
+          this.checkToken();
+          this.token += character;
+          // makes sure column position is accurate for operator
+          this.trace.column[0] = this.virtual_position;
+          this.checkToken(1);
+        } else {
+          // if no conditions met, keep building token
+          this.token += character;
+        }
+        break;
+      }
+
+      case Mode.String: {
```

Comment:
Currently, there's no section for parsing `\n` in a string -- this should be added. There are a LOT of cases when `invalid_string` will fire, even when unintentional. I think this is caused by the eol token -- See comment on 36.

Overall I think strings need lots of work, and we should be carefully simplifying them. This was by far the most complex system in the legacy lexer, and that was for a reason -- there's a lot of stuff to check. We'll want to include linebreaks, spaces, escapes, and various other items within strings... this should be thoroughly tested before implementing it.

Reply:
Lexing of \n is definitely a fix that needs to be implemented, but general understanding of strings by the lexer is unneeded. The lexer should feed in the string data, and leave understanding of special escaped characters to later sequences, such as parser or even rudimentary code generation. Escape sequences are not something easy to perfect; I'll go ahead and add naive escape sequence understanding so they're retained in strings 👍.

Reply:
Agreed. I wasn't saying we should actively understand escaped sections in a string, leave that to the parser, but we should definitely include them in the string token. As long as escapes are retained in string tokens, we should be good to go from the lexer's perspective. Side note: Make sure to remove the backslashes. For example, `"this is \"escaped\""` should create a string token with a value of `this is "escaped"`. In this version of the PR, it gets buggy and will break up the string token. I actually think this may be the same bug as [this](https://github.com/mtl-lang/MTL/pull/1#discussion_r999928079) comment, or at the least they're related.

(Based on your comment [here](https://github.com/mtl-lang/MTL/pull/1#discussion_r1001250075) I think this may already be fixed. Feel free to resolve once tested)